### ISSUE TITLE
Linear Regression 의 cost 최소화의 TensorFlow 구현

### ISSUE THREAD

**Issue by NahahaNahaha** [NahahaNahaha/tensorflow#5]
import tensorflow as tf
import matplotlib.pyplot as plt

X = [1, 2, 3]
Y = [1, 2, 3]

W = tf.placeholder(tf.float32)

hypothesis = X * W

cost = tf.reduce_mean(tf.square(hypothesis - Y))

sess = tf.Session()

sess.run(tf.global_variables_initializer())
여기까지는 지난번 코드와 같다. 
W_val = []
cost_val = [] 

for i in range(-30, 50):
   feed_W = i * 0.1
   curr_cost, curr_W = sess.run([cost, W], feed_dict={W: feed_W})
   W_val.append(curr_W)
   cost_val.append(curr_cost) 
i가 -30에서 50까지 반복되는데 0.1을 곱하므로 -3~5까지 feed_W에 저장.. 

append 명령어는 데이터들을 저장시키는 명령어라고 한다. W_val과 cost_val 에 각각 curr_W(W 값들) 과 curr_cost(그에 대응하는 cost 값들) 을 넣어준다.  

plt.plot(W_val, cost_val) 
plt.show() 
그렇게 해서 그리면 다음과 같이 생성된다,. 
<img width="643" alt="2018-03-30 2 29 42" src="https://user-images.githubusercontent.com/26895961/38126002-b19becb6-3428-11e8-8727-863b8163c477.png">
이차함수 모양으로 이상적인 W값 1을 가진다. 

이제 우리가 배운 Gradient Descent를 통한 최소화 코드를 넣어보면 

import tensorflow as tf
x_data = [1, 2, 3]
y_data = [1, 2, 3]

W = tf.Variable(tf.random_normal([1]), name='weight')
X = tf.placeholder(tf.float32)
Y = tf.placeholder(tf.float32)

hypothesis = X * W

cost = tf.reduce_sum(tf.square(hypothesis - Y))
지난 시간에 배웠던 코드이다. 

learning_rate = 0.1. //cost 를 최소화 시키는 W에 관한 식에서 상수 알파를 0.1로 설정
gradient = tf.reduce_mean((W * X - Y) * X) 
descent = W - learning_rate * gradient. //W를 점차적으로 최소화 시키는 식
update = W.assign(descent) //위 식을 계속 실행시키기 위해서 update라는 변수로 지정

sess = tf.Session()
sess.run(tf.global_variables_initializer())
for step in range(21):
   sess.run(update, feed_dict={X: x_data, Y: y_data})
   print(step, sess.run(cost, feed_dict={X: x_data, Y: y_data}), sess.run(W)) 
이를 실행시키면 다음과 같다. 
(0, 14.255027, array([-0.009067], dtype=float32))
array([0.4618309], dtype=float32)
(1, 4.0547633, array([0.4618309], dtype=float32))
array([0.71297646], dtype=float32)
(2, 1.1533551, array([0.71297646], dtype=float32))
array([0.8469208], dtype=float32)
(3, 0.32806537, array([0.8469208], dtype=float32))
array([0.91835773], dtype=float32)
(4, 0.093316495, array([0.91835773], dtype=float32))
array([0.9564575], dtype=float32)
(5, 0.026543327, array([0.9564575], dtype=float32))
array([0.9767773], dtype=float32)
(6, 0.0075501036, array([0.9767773], dtype=float32))
array([0.9876146], dtype=float32)
(7, 0.002147588, array([0.9876146], dtype=float32))
array([0.99339443], dtype=float32)
(8, 0.0006108666, array([0.99339443], dtype=float32))
array([0.996477], dtype=float32)
(9, 0.00017376315, array([0.996477], dtype=float32))
array([0.9981211], dtype=float32)
(10, 4.942395e-05, array([0.9981211], dtype=float32))
array([0.9989979], dtype=float32)
(11, 1.4058112e-05, array([0.9989979], dtype=float32))
array([0.9994656], dtype=float32)
(12, 3.9980127e-06, array([0.9994656], dtype=float32))
array([0.999715], dtype=float32)
(13, 1.1371808e-06, array([0.999715], dtype=float32))
array([0.99984795], dtype=float32)
(14, 3.236206e-07, array([0.99984795], dtype=float32))
array([0.9999189], dtype=float32)
(15, 9.215974e-08, array([0.9999189], dtype=float32))
array([0.9999567], dtype=float32)
(16, 2.6184765e-08, array([0.9999567], dtype=float32))
array([0.99997693], dtype=float32)
(17, 7.4409634e-09, array([0.99997693], dtype=float32))
array([0.9999877], dtype=float32)
(18, 2.1019133e-09, array([0.9999877], dtype=float32))
array([0.99999344], dtype=float32)
(19, 5.971543e-10, array([0.99999344], dtype=float32))
array([0.9999965], dtype=float32)
(20, 1.7188384e-10, array([0.9999965], dtype=float32)) 
차례대로 step, cost, W값이다. 마찬가지로 cost는 0에 W는 1에 가까워지는 것을 볼 수 있다. 
..
..
..
hypothesis = X * W
cost = tf.reduce_mean(tf.square(hypothesis - Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.1)
train = optimizer.minimize(cost)

지난 시간에 원리를 몰랐던 함수 GradientDescentOptimizer가 무엇인지 밝혀졌다. 예제 케이스의 경우 함수가 간단하여 미분하기 쉬워 직접해서 식을 넣는게 가능했지만 복잡한 함수의 경우 그것이 불가능하다. 그래서 쓰는 함수가 이 GradientDescentOptimizer 함수인 것이다. 자동으로 미분해서 상수 알파(learning_rate)를 곱한뒤 빼준다. 
그 후 minimize 라는 것을 통해 최소화를 시켜준다. 

import tensorflow as tf
X = [1, 2, 3]
Y = [1, 2, 3]
W = tf.Variable(5.) //W의 초기값 5
hypothesis = X * W
gradient = tf.reduce_mean((W * X - Y) * X) * 2
cost = tf.reduce_mean(tf.square(hypothesis - Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.01)

gvs = optimizer.compute_gradients(cost, [W])
apply_gradients = optimizer.apply_gradients(gvs)

sess = tf.Session()
sess.run(tf.global_variables_initializer())

for step in range(100):
   print(step, sess.run([gradient, W, gvs]))
   sess.run(apply_gradients)
0, [37.333332, 5.0, [(37.333336, 5.0)]])
(1, [33.84889, 4.6266665, [(33.84889, 4.6266665)]])
(2, [30.689657, 4.2881775, [(30.689657, 4.2881775)]])
(3, [27.825287, 3.9812808, [(27.825287, 3.9812808)]])
(4, [25.228262, 3.703028, [(25.228264, 3.703028)]])
(5, [22.873621, 3.4507453, [(22.873623, 3.4507453)]])
(6, [20.738752, 3.2220092, [(20.738752, 3.2220092)]])
(7, [18.803137, 3.0146217, [(18.803137, 3.0146217)]])
(8, [17.048176, 2.8265903, [(17.048176, 2.8265903)]])
(9, [15.457013, 2.6561086, [(15.457014, 2.6561086)]])
(10, [14.014359, 2.5015385, [(14.01436, 2.5015385)]])
(11, [12.706352, 2.361395, [(12.706352, 2.361395)]])
(12, [11.520427, 2.2343314, [(11.520427, 2.2343314)]])
(13, [10.445186, 2.119127, [(10.445186, 2.119127)]])
(14, [9.470302, 2.0146751, [(9.470302, 2.0146751)]])
(15, [8.586407, 1.9199722, [(8.586407, 1.9199722)]])
(16, [7.785009, 1.8341081, [(7.785009, 1.8341081)]])
(17, [7.0584083, 1.756258, [(7.0584083, 1.756258)]])
(18, [6.399624, 1.685674, [(6.399624, 1.685674)]])
(19, [5.8023257, 1.6216778, [(5.8023257, 1.6216778)]])
(20, [5.260776, 1.5636545, [(5.260776, 1.5636545)]])
(21, [4.7697697, 1.5110468, [(4.7697697, 1.5110468)]])
(22, [4.324591, 1.4633491, [(4.324591, 1.4633491)]])
(23, [3.9209633, 1.4201032, [(3.9209635, 1.4201032)]])
(24, [3.5550067, 1.3808936, [(3.5550067, 1.3808936)]])
(25, [3.2232056, 1.3453435, [(3.2232056, 1.3453435)]])
(26, [2.9223735, 1.3131114, [(2.9223738, 1.3131114)]])
(27, [2.6496189, 1.2838877, [(2.6496186, 1.2838877)]])
(28, [2.4023216, 1.2573916, [(2.4023218, 1.2573916)]])
(29, [2.178105, 1.2333684, [(2.178105, 1.2333684)]])
(30, [1.9748148, 1.2115873, [(1.9748147, 1.2115873)]])
(31, [1.7904993, 1.1918392, [(1.7904994, 1.1918392)]])
(32, [1.623386, 1.1739342, [(1.6233861, 1.1739342)]])
(33, [1.4718695, 1.1577003, [(1.4718695, 1.1577003)]])
(34, [1.3344955, 1.1429816, [(1.3344957, 1.1429816)]])
(35, [1.2099417, 1.1296366, [(1.2099419, 1.1296366)]])
(36, [1.0970144, 1.1175373, [(1.0970144, 1.1175373)]])
(37, [0.9946267, 1.1065671, [(0.9946267, 1.1065671)]])
(38, [0.90179497, 1.0966209, [(0.901795, 1.0966209)]])
(39, [0.8176275, 1.087603, [(0.81762755, 1.087603)]])
(40, [0.7413151, 1.0794266, [(0.7413151, 1.0794266)]])
(41, [0.67212623, 1.0720135, [(0.6721263, 1.0720135)]])
(42, [0.609394, 1.0652922, [(0.609394, 1.0652922)]])
(43, [0.5525169, 1.0591983, [(0.5525169, 1.0591983)]])
(44, [0.50094914, 1.0536731, [(0.50094914, 1.0536731)]])
(45, [0.45419374, 1.0486636, [(0.45419377, 1.0486636)]])
(46, [0.41180158, 1.0441216, [(0.41180158, 1.0441216)]])
(47, [0.37336722, 1.0400037, [(0.37336725, 1.0400037)]])
(48, [0.33851996, 1.03627, [(0.33852, 1.03627)]])
(49, [0.30692515, 1.0328848, [(0.30692515, 1.0328848)]])
(50, [0.27827826, 1.0298156, [(0.2782783, 1.0298156)]])
(51, [0.25230527, 1.0270327, [(0.25230527, 1.0270327)]])
(52, [0.2287569, 1.0245097, [(0.2287569, 1.0245097)]])
(53, [0.20740573, 1.022222, [(0.20740573, 1.022222)]])
(54, [0.18804836, 1.020148, [(0.18804836, 1.020148)]])
(55, [0.17049654, 1.0182675, [(0.17049655, 1.0182675)]])
(56, [0.15458433, 1.0165626, [(0.15458435, 1.0165626)]])
(57, [0.14015675, 1.0150168, [(0.14015675, 1.0150168)]])
(58, [0.12707591, 1.0136153, [(0.12707591, 1.0136153)]])
(59, [0.11521538, 1.0123445, [(0.11521538, 1.0123445)]])
(60, [0.10446167, 1.0111923, [(0.10446167, 1.0111923)]])
(61, [0.09471202, 1.0101477, [(0.09471202, 1.0101477)]])
(62, [0.08587202, 1.0092006, [(0.08587202, 1.0092006)]])
(63, [0.07785805, 1.0083419, [(0.07785805, 1.0083419)]])
(64, [0.07059129, 1.0075634, [(0.07059129, 1.0075634)]])
(65, [0.06400236, 1.0068574, [(0.06400236, 1.0068574)]])
(66, [0.05802846, 1.0062174, [(0.05802846, 1.0062174)]])
(67, [0.052612226, 1.005637, [(0.052612226, 1.005637)]])
(68, [0.047702473, 1.005111, [(0.047702473, 1.005111)]])
(69, [0.043249767, 1.0046339, [(0.043249767, 1.0046339)]])
(70, [0.03921318, 1.0042014, [(0.03921318, 1.0042014)]])
(71, [0.035553534, 1.0038093, [(0.035553537, 1.0038093)]])
(72, [0.032236177, 1.0034539, [(0.03223618, 1.0034539)]])
(73, [0.029227654, 1.0031315, [(0.029227655, 1.0031315)]])
(74, [0.02649951, 1.0028392, [(0.02649951, 1.0028392)]])
(75, [0.024025917, 1.0025742, [(0.024025917, 1.0025742)]])
(76, [0.021783749, 1.002334, [(0.02178375, 1.002334)]])
(77, [0.01975123, 1.0021162, [(0.019751232, 1.0021162)]])
(78, [0.017907381, 1.0019187, [(0.017907381, 1.0019187)]])
(79, [0.016236702, 1.0017396, [(0.016236704, 1.0017396)]])
(80, [0.014720838, 1.0015773, [(0.014720838, 1.0015773)]])
(81, [0.01334699, 1.00143, [(0.013346991, 1.00143)]])
(82, [0.012100856, 1.0012965, [(0.012100856, 1.0012965)]])
(83, [0.010971785, 1.0011755, [(0.010971785, 1.0011755)]])
(84, [0.0099481745, 1.0010659, [(0.009948175, 1.0010659)]])
(85, [0.009018898, 1.0009663, [(0.009018898, 1.0009663)]])
(86, [0.008176883, 1.0008761, [(0.008176884, 1.0008761)]])
(87, [0.007413149, 1.0007943, [(0.007413149, 1.0007943)]])
(88, [0.006721576, 1.0007201, [(0.006721576, 1.0007201)]])
(89, [0.0060940585, 1.0006529, [(0.0060940585, 1.0006529)]])
(90, [0.005525271, 1.000592, [(0.0055252714, 1.000592)]])
(91, [0.0050098896, 1.0005368, [(0.0050098896, 1.0005368)]])
(92, [0.004542589, 1.0004867, [(0.004542589, 1.0004867)]])
(93, [0.0041189194, 1.0004413, [(0.0041189194, 1.0004413)]])
(94, [0.0037339528, 1.0004001, [(0.003733953, 1.0004001)]])
(95, [0.0033854644, 1.0003628, [(0.0033854644, 1.0003628)]])
(96, [0.0030694802, 1.0003289, [(0.0030694804, 1.0003289)]])
(97, [0.0027837753, 1.0002983, [(0.0027837753, 1.0002983)]])
(98, [0.0025234222, 1.0002704, [(0.0025234222, 1.0002704)]])
(99, [0.0022875469, 1.0002451, [(0.0022875469, 1.0002451)]])

그 결과값이다. 차례대로 step, (gradient 일 때의 cost, W) 그리고 (GradientDescentOptimizer를 썼을 때의 cost, W) 이다. 

거의 같은 값을 보인다. ( 조금의 차이는 단지 소수점계산 때문의 차이이다.) 

마지막으로 

import tensorflow as tf
X = [1, 2, 3]
Y = [1, 2, 3]
W = tf.Variable(5.)
hypothesis = X * W
gradient = tf.reduce_mean((W * X - Y) * X) * 2
cost = tf.reduce_mean(tf.square(hypothesis - Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.01)

gvs = optimizer.compute_gradients(cost, [W])
apply_gradients = optimizer.apply_gradients(gvs)

sess = tf.Session()
sess.run(tf.global_variables_initializer())

for step in range(100):
   print(step, sess.run([gradient, W, gvs]))
   sess.run(apply_gradients) 

이렇게 gvs라는 변수를 이용해서 최소화 시킨 gradient 값을 지정시킨후 원하는대로 조작해서 함수를 돌릴 수도 있다.(위에서는 그대로 gvs 값을 넣었기 때문에 변화가 없다. )
